### PR TITLE
Implemented new event `renderable:renderLifecycle:anyEvent`  to make it easier to listen to renderable events

### LINKED ISSUE #359: Emit events and the start and end of rendering for every `Renderable`, send to root circuit
Two main events:

- `renderable:renderLifecycle:<phase name>:start`
- `renderable:renderLifecycle:<phase name>:end`

e.g.

- `renderable:renderLifecycle:doInitialPcbComponentRender:start`
- `renderable:renderLifecycle:doInitialPcbComponentRender:end`

(the root circuit is the circuit defined in `Circuit.ts`, commonly used as `this.root`, all events from renderable are re-emitted)

The event payloads should include:

```js
{ renderId }
```

Create the utility event `"renderable:renderLifecycle:anyEvent"` to make it easier to listen to renderable events.

Make sure to test by recording all the events that are emitted using `circuit.on("renderable:anyEvent")

/bounty $30

### PR DESCRIPTION
**Pull request by anas-sarkez**
/claim #359
/closes #359

**example of how it works** 

Screenshot 2024-11-27 at 1 48 43 PM

### CODE REVIEW

File: lib/Circuit.ts [hunk 1]
```diff
@@ -6,7 +6,11 @@ import { isValidElement, type ReactElement } from "react"
 import { createInstanceFromReactElement } from "./fiber/create-instance-from-react-element"
 import { identity, type Matrix } from "transformation-matrix"
 
-type RootCircuitEventName = "asyncEffectComplete"
+type RootCircuitEventName =
+  | "asyncEffectComplete"
+  | "renderable:renderLifecycle:anyEvent"
+  | `renderable:renderLifecycle:${string}:start`
```
Comment (seveibar) [inline lib/Circuit.ts:12]: Can we make this slightly better:

```suggestion
  | `renderable:renderLifecycle:${RenderPhase}:start`
```

File: lib/components/base-components/Renderable.ts [hunk 2]
```diff
@@ -179,11 +179,27 @@ export abstract class Renderable implements IRenderable {
...
+      ;(this as any).root
+        ?.emit(
+          eventName,
+          eventPayload,
+        )(this as any)
+        .root?.emit("renderable:renderLifecycle:anyEvent", {
+          ...eventPayload,
+          eventName,
+        })
+    }
```
Comment (seveibar) [inline lib/components/base-components/Renderable.ts:194]: don't use an inline function here- make it a proper class method `this._emitRenderLifecycleEvent(renderPhase, "start")`

File: lib/components/base-components/Renderable.ts [hunk 3]
```diff
@@ -197,18 +213,23 @@ export abstract class Renderable implements IRenderable {
       if (hasIncompleteEffects) return
     }
 
+    // Handle start
+    emitRenderEvent("start")
```
Comment (seveibar) [inline lib/components/base-components/Renderable.ts:217]: Similar to my other comment, i don't like how the phase is implicit here, I'd prefer if the phase was explicitly specified

File: lib/components/base-components/Renderable.ts [hunk 4]
```diff
@@ -179,11 +179,27 @@ export abstract class Renderable implements IRenderable {
...
+      const eventPayload = { renderId: this._renderId }
+      const eventName = `renderable:renderLifecycle:${phase}:${eventType}`
+      ;(this as any).root
+        ?.emit(
+          eventName,
+          eventPayload,
+        )(this as any)
+        .root?.emit("renderable:renderLifecycle:anyEvent", {
+          ...eventPayload,
+          eventName,
```
Comment (seveibar) [inline lib/components/base-components/Renderable.ts:192]: The Event standard specifies "type" as the key: https://developer.mozilla.org/en-US/docs/Web/API/Event

Can you also add `componentDisplayName: this.getString()` to help us identify the component being logged.

```suggestion
          type: eventName,
```

**Comment by seveibar** [comment]
nice work, a couple tweaks and we're good to go

**Comment by anas-sarkez** [comment]
@seveibar 
check these updates 
Screenshot 2024-11-27 at 3 49 34 PM

**Review by seveibar — APPROVED** [review]
GOOD SHOW

**Pull request merged**